cmd/mt-control-server/handlers.go

```diff
@@ -0,0 +1,339 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/gocql/gocql"
+	"github.com/grafana/globalconf"
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/api/response"
+	"github.com/grafana/metrictank/cmd/mt-control-server/controlmodels"
+	"github.com/grafana/metrictank/expr/tagquery"
+	"github.com/grafana/metrictank/idx"
+	"github.com/grafana/metrictank/schema"
+	"github.com/grafana/metrictank/schema/msg"
+	"github.com/grafana/metrictank/util"
+	log "github.com/sirupsen/logrus"
+	"github.com/tinylib/msgp/msgp"
+	"gopkg.in/macaron.v1"
+)
+
+var metrictankUrl string
+var maxDefsPerMsg int
+var indexConcurrency int
+
+func ConfigHandlers() {
+	FlagSet := flag.NewFlagSet("handlers", flag.ExitOnError)
+
+	FlagSet.StringVar(&metrictankUrl, "metrictank-url", "http://metrictank:6060", "URL for MT cluster")
+	FlagSet.IntVar(&maxDefsPerMsg, "max-defs-per-msg", 1000, "Maximum defs per control message")
+	FlagSet.IntVar(&indexConcurrency, "index-concurrency", 5, "Concurrent queries to run against cassandra (per request)")
+
+	globalconf.Register("handlers", FlagSet, flag.ExitOnError)
+}
+
+func makeLogId() string {
+	logId := "uninit"
+	uuid, err := gocql.RandomUUID()
+	if err == nil {
+		logId = uuid.String()
+	}
+	return logId
+}
+
+func makeMessage(partition int32, op schema.Operation, cm *schema.ControlMsg) (*sarama.ProducerMessage, error) {
+	cm.Op = op
+	var b bytes.Buffer
+	b.WriteByte(byte(msg.FormatIndexControlMessage))
+	w := msgp.NewWriterSize(&b, 300)
+	err := cm.EncodeMsg(w)
+	if err != nil {
+		return nil, err
+	}
+	w.Flush()
+
+	return &sarama.ProducerMessage{
+		Topic:     topic,
+		Value:     sarama.ByteEncoder(b.Bytes()),
+		Partition: partition,
+	}, nil
+}
+
+func appStatus(ctx *macaron.Context) {
+	ctx.PlainText(200, []byte("OK"))
+}
+
+func tagsDelByQuery(ctx *macaron.Context, request controlmodels.IndexDelByQueryReq) {
+	logId := makeLogId()
+
+	log.Infof("Received tagsDelByQuery: logId = %s, req = %v", logId, request)
+
+	// Translate to FindSeries
+	findSeries := models.GraphiteTagFindSeries{
+		Expr:   request.Expr,
+		Limit:  request.Limit,
+		To:     request.OlderThan,
+		Meta:   true,
+		Format: "full-json",
+	}
+	reqBody, err := json.Marshal(findSeries)
+	if err != nil {
+		response.Write(ctx, response.WrapError(err))
+		return
+	}
+	// send to query address
+	resp, err := http.Post(metrictankUrl+"/tags/findSeries", "application/json", bytes.NewBuffer(reqBody))
+	if err != nil {
+		response.Write(ctx, response.WrapError(err))
+		return
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		// Read in body so that the connection can be reused
+		io.Copy(ioutil.Discard, resp.Body)
+		response.Write(ctx, response.WrapError(fmt.Errorf("Request to cluster failed with status = %d", resp.StatusCode)))
```

Comment:
we probably shouldn't use the same response code in our responses.
for example, upstream may return 404, that doesn't mean the response from mt-control-server should also be 404.

Perhaps 502 is a better choice, any time upstream has an issue?